src/appleseed/renderer/modeling/frame/frame.cpp

```diff
@@ -462,28 +490,465 @@ void Frame::denoise(
 
 namespace
 {
-    void create_parent_directories(const bf::path& file_path)
+
+    typedef vector<tuple<string, CanvasProperties, ImageAttributes>> CheckpointProperties;
+
+    //
+    // Interface used to save the rendering buffer in checkpoints.
+    //
+
+    class ShadingBufferCanvas
+      : public ICanvas
+    {
+      public:
+        ShadingBufferCanvas(
+            const Frame&                                frame,
+            PermanentShadingResultFrameBufferFactory*   buffer_factory)
```

Comment:
Do you really need a `PermanentShadingResultFrameBufferFactory`? What for?

Reply:
Yes because `EphemeralShadingResultFrameBufferFactory` always create a new tile. But here I want the tiles that was already created and filled.

Reply:
If the `ShadingBufferCanvas` would work with any `IShadingResultFrameBufferFactory`, then you wouldn't need casts or any special cases anywhere.